### PR TITLE
perf: retry stack analysis service

### PR DESCRIPTION
**Pull request by IlonaShishov**
PR addresses an error occurring when requesting a dependency analysis report - status code 504 (which indicates that a server acting as a gateway was not able to get a response from the upstream server that it needs to complete the request).
In an effort to try and reduce the likeliness of this error occurring I have included an Exponential Backoff retry to the POST and GET services for stack analysis.

### CODE REVIEW

File: src/stackAnalysisService.ts [hunk 1]
```diff
@@ -3,18 +3,27 @@
...
-          reject(err);
+          if (retryCount < MAX_RETRIES) {
+            retryCount++;
+            invokeExponentialBackoff(retryCount, getRequestWithExponentialBackoff, resolve, reject);
+          } else {
+            reject(err);
+          }
         } else {
           if (
             httpResponse.statusCode === 200 ||
```
Comment (ilan-pinto) [inline src/stackAnalysisService.ts:29]: is there any chance backend will send back 202 or 201?

File: src/stackAnalysisService.ts [hunk 2]
```diff
@@ -23,41 +32,50 @@ export module stackAnalysisServices {
...
-              httpResponse.statusCode
-            }`;
+            errorMsg = `Failed :: ${httpResponse.statusMessage}, Status: ${httpResponse.statusCode
+              }`;
             reject(errorMsg);
+          } else if (retryCount < MAX_RETRIES) {
+            retryCount++;
+            invokeExponentialBackoff(retryCount, getRequestWithExponentialBackoff, resolve, reject);
           } else if (
             httpResponse.statusCode === 429 ||
```
Comment (ilan-pinto) [inline src/stackAnalysisService.ts:42]: What happens for 5XX response? 
Why not retry for any non-2XX response?

File: test/stackAnalysisService.test.ts [hunk 3]
```diff
@@ -112,20 +113,21 @@ suite('stacknalysis Services', () => {
...
       .stub(request, 'post')
       .yields(null, { statusCode: 500 });
-    stackAnalysisServices.postStackAnalysisService(options, context);
+    stackAnalysisServices.postStackAnalysisService(options, context, retryCount);
     expect(stubRequestPost).callCount(1);
     expect(spyClearContextInfo).callCount(1);
   });
 
   test('postStackAnalysisService should return error', () => {
+    let retryCount = 999;
```
Comment (ilan-pinto) [inline test/stackAnalysisService.test.ts:122]: Why 999 retries?

**Review by ilan-pinto — COMMENTED** [review]
In the PR desc, you mentioned that issue was caused by error 504 in the code you don't address it

File: test/stackAnalysisService.test.ts (hunk 3 above)
Comment (IlonaShishov) [reply test/stackAnalysisService.test.ts:122]: the goal in this test is to skip the retry mechanism and only test the functionality itself. 
in order to surpass any logical MAX_RETRIES value, so this statement `else if (retryCount < MAX_RETRIES)` will return false, retry must be a very high/unrealistic number. 
so i set it to 999

File: src/stackAnalysisService.ts (hunk 1 above)
Comment (IlonaShishov) [reply src/stackAnalysisService.ts:29]: for status code 202, that is already included in the if statement 
```
if (
            httpResponse.statusCode === 200 ||
            httpResponse.statusCode === 202
    )
```

as for status code 201, HTTP status code 201 is specifically used to indicate that a POST, PUT, or PATCH request has been successful and that a new resource has been created. A GET request would not likely return a status code of 201. 
according to this, the method getStackAnalysisService does not require status 201 to be considered.
On the other hand, method postStackAnalysisService could take status code 201 into consideration.

But judging by the fact that the creators of this code did not include status code 201 to begin with, can we be confident that they considered their API to never return such a status code?

Comment (ilan-pinto) [reply src/stackAnalysisService.ts:29]: Agreed.

File: src/stackAnalysisService.ts [hunk 4]
```diff
@@ -68,44 +86,54 @@ export module stackAnalysisServices {
...
+            reject(errorMsg);
           } else if (httpResponse.statusCode === 401) {
             clearContextInfo(context);
-            errorMsg = `Failed :: ${httpResponse.statusMessage}, Status: ${
-              httpResponse.statusCode
-            }`;
+            errorMsg = `Failed :: ${httpResponse.statusMessage}, Status: ${httpResponse.statusCode
+              }`;
             reject(errorMsg);
+          } else if (retryCount < MAX_RETRIES) {
```
Comment (ilan-pinto) [inline src/stackAnalysisService.ts:102]: I think you should move the "retry if" below next if. Otherwise, you will retry while getting 403

Comment (IlonaShishov) [reply src/stackAnalysisService.ts:102]: The creators of this code have categorized status code 403 and 429 (in one if statement) as a `server busy` error, message `Service is currently busy to process your request for analysis, please try again in few minutes`.
Im not sure why, but i assume the reasoning comes from the behavior of the server in these cases. 

According to this categorization, shouldn't both status codes in this case be retried?

File: test/stackAnalysisService.test.ts (hunk 3 above)
Comment (IlonaShishov) [reply test/stackAnalysisService.test.ts]: I agree, thanks for the catch!
Ill take care of it using the Numbers.MAX_VALUE property

File: src/stackAnalysisService.ts (hunk 4 above)
Comment (ilan-pinto) [reply src/stackAnalysisService.ts:102]: yes

**Review by jparsai — APPROVED** [review]

**Pull request merged**